File: lintcode/cpp/Lowest_Common_Ancestor.cpp
/**
 * Solution: find the first element which has non empty left and right returned
 *  - top->down: search the element, and return the node has A or B
 *  - bottom->up: based on the searched result to return the FIRST node with non-empty left or right return
 * 
 *    5 <- A
 *     \
 *      6 <- B
 *  if A or B are the LCA itself, it works as well.
 */
class Solution {
public:
    /**
     * @param root: The root of the binary search tree.
     * @param A and B: two nodes in a Binary.
     * @return: Return the least common ancestor(LCA) of the two nodes.
     */
    TreeNode *lowestCommonAncestor(TreeNode *root, TreeNode *A, TreeNode *B) {
        // write your code here
        if(!root) return nullptr;
        if(root == A || root == B) return root;
        
        TreeNode *l = lowestCommonAncestor(root->left, A, B);
        TreeNode *r = lowestCommonAncestor(root->right, A, B);
        
        if(!l) return r;
        if(!r) return l;
        return root;
    }
};

File: lintcode/java/LowestCommonAncestor.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param root: The root of the binary search tree.
     * @param A and B: two nodes in a Binary.
     * @return: Return the least common ancestor(LCA) of the two nodes.
     */
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode A, TreeNode B) {
        if(root == null || A == null || B == null) return null;
        if(root == A || root == B) return root;
        
        TreeNode l = lowestCommonAncestor(root.left, A, B);
        TreeNode r = lowestCommonAncestor(root.right, A, B);
        
        if(l == null) return r;
        if(r == null) return l;
        
        // first node with l != null && r != null
        // otherwise, the LCA node will be returned all the way to the top
        // NOTES:
        // 1. left or right subtree returns null
        // 2. left and right subtree does not return null
        return root;
    }
}
